### PR TITLE
Add support for 'SELECT TOP <n>' as an alternative syntax for LIMIT.

### PR DESCRIPTION
**Pull request by ptallada**
Hi,

I think I've got it ;)

### CODE REVIEW

**Comment by ptallada** [comment]
I don't understand the lint error :/

File: sqlglot/parser.py [hunk 1]
```diff
@@ -795,13 +795,19 @@ def _parse_select(self):
             )
 
         if isinstance(this, exp.Subquery) or self._match(TokenType.SELECT):
+            hint=self._parse_hint()
+            _=self._match(TokenType.ALL)
```
Comment (barakalon) [inline sqlglot/parser.py:799]: Should we be matching ALL _or_ DISTINCT?

**Review by barakalon — COMMENTED** [review]
Nice!

One primary concern: This parses TOP, but it can't generate TOP. 

Also: add tests. If you add one [here](https://github.com/tobymao/sqlglot/blob/main/tests/fixtures/identity.sql), I think you'll see the above problem.

Happy to talk through ideas on how we can handle that!

**Comment by barakalon** [comment]
That's not a lint error - it's a format error. Run `./format_code.sh`.

**Comment by tobymao** [comment]
it may not be necessary  to generate top

**Comment by barakalon** [comment]
Not generating TOP feels like a bandaid for parsing SQLServer syntax. I think a true _transpiler_ would be able to go to _and_ from different dialects.

**Comment by barakalon** [comment]
Yeah, offset makes this even messier

File: sqlglot/parser.py (hunk 1 above)
Comment (ptallada) [reply sqlglot/parser.py]: Yes, it is one or the other, I'll try to fix that

**Comment by ptallada** [comment]
> Nice!
> 
> One primary concern: This parses TOP, but it can't generate TOP.
Yes, although for me it is not a requirement...

> Also: add tests. If you add one [here](https://github.com/tobymao/sqlglot/blob/main/tests/fixtures/identity.sql), I think you'll see the above problem.
On it :)

> Happy to talk through ideas on how we can handle that!
If it is not very difficult, I can invest a pair of hours on it :)

**Comment by ptallada** [comment]
> https://stackoverflow.com/questions/603724/how-to-implement-limit-with-sql-server
> 
> offset fetch

WTF, I was not aware of this mess. My requirement is to transpile ADQL (not MSSQL), which seems to be based on the former (and easier) MSSQL syntax.

File: sqlglot/parser.py [hunk 2]
```diff
@@ -811,7 +820,7 @@ def _parse_select(self):
                     "having": self._parse_having(),
                     "qualify": self._parse_qualify(),
                     "order": self._parse_order(),
-                    "limit": self._parse_limit(),
+                    "limit": limit if limit else self._parse_limit(),
```
Comment (tobymao) [inline sqlglot/parser.py:823]: limit or self._parse_limit()

File: sqlglot/parser.py [hunk 3]
```diff
@@ -1039,6 +1048,11 @@ def _parse_limit(self):
             return None
         return self.expression(exp.Limit, this=self._parse_number())
 
+    def _parse_top(self):
```
Comment (tobymao) [inline sqlglot/parser.py:1051]: let's just delete this and add an arg to _parse_limit which is 

def _parse_limit(self, top=False):
self._match(TokenType.TOP if top else TokenType.LIMIT)

File: sqlglot/tokens.py [hunk 4]
```diff
@@ -373,6 +374,7 @@ class Tokenizer:
         "TEMPORARY": TokenType.TEMPORARY,
         "THEN": TokenType.THEN,
         "TIME": TokenType.TIME,
+        "TOP": TokenType.TOP,
```
Comment (tobymao) [inline sqlglot/tokens.py:377]: and TokenType.TOP to the list of id_vars in parser.py

**Comment by tobymao** [comment]
> Ok, I am a bit lost on how to implement TOP generation, any hints?

you probably need to add argument to Generator to support TOP generation, as i said before, i'm fine if we don't add it

**Review by tobymao — APPROVED** [review]

**Pull request merged**